R/forecast.R

```diff
@@ -28,55 +42,31 @@
 #' - `forecast_point` for point forecasts
 #' - `forecast_sample` for sample-based forecasts
 #' - `forecast_quantile` for quantile-based forecasts
-#' @export
 #' @keywords check-forecasts
 #' @examples
-#' as_forecast(example_binary)
-#' as_forecast(
+#' as_forecast_binary(example_binary)
+#' as_forecast_quantile(
 #'   example_quantile,
 #'   forecast_unit = c("model", "target_type", "target_end_date",
 #'                     "horizon", "location")
 #' )
-as_forecast <- function(data, ...) {
-  UseMethod("as_forecast")
-}
+#' @name as_forecast
+NULL
 
-#' @rdname as_forecast
-#' @param forecast_unit (optional) Name of the columns in `data` (after
-#'   any renaming of columns done by `as_forecast()`) that denote the unit of a
-#'   single forecast. See [get_forecast_unit()] for details.
-#'   If `NULL` (the default), all columns that are not required columns are
-#'   assumed to form the unit of a single forecast. If specified, all columns
-#'   that are not part of the forecast unit (or required columns) will be removed.
-#' @param forecast_type (optional) The forecast type you expect the forecasts
-#'   to have. If the forecast type as determined by `scoringutils` based on the
-#'   input does not match this, an error will be thrown. If `NULL` (the
-#'   default), the forecast type will be inferred from the data.
-#' @param observed (optional) Name of the column in `data` that contains the
-#'   observed values. This column will be renamed to "observed".
-#' @param predicted (optional) Name of the column in `data` that contains the
-#'   predicted values. This column will be renamed to "predicted".
-#' @param model (optional) Name of the column in `data` that contains the names
-#'   of the models/forecasters that generated the predicted values.
-#'   This column will be renamed to "model".
-#' @param quantile_level (optional) Name of the column in `data` that contains
-#'   the quantile level of the predicted values. This column will be renamed to
-#'   "quantile_level". Only applicable to quantile-based forecasts.
-#' @param sample_id (optional) Name of the column in `data` that contains the
-#'   sample id. This column will be renamed to "sample_id". Only applicable to
-#'   sample-based forecasts.
-#' @export
-#' @importFrom cli cli_warn
-as_forecast.default <- function(data,
+
+#' Common functionality for `as_forecast_...` functions
```

Comment:
if we expect people to build their own forecast classes and we think this might help them with that then surely this should be exported? The other option we could do here is keep this as the as_forecast() method and this would give the class hierarchy (everything would have aa specific class and a forecast class) if as_forecast() was called inside all the other methods?

You could also use as_forecast as the user to check you have the generic bits you need before doing the full class conversion (not sure this is a good idea)

Reply:
Hm. At the moment this helper function doesn't actually return a `forecast` object (meaning that the name might be a bit misleading). That's why I left it internal (also given that it's currently really hard to create your own forecast class). 
I suggest the following: I opened up an issue to discuss/flesh out the way in which users are expected to add a new forecast class. This likely also requires some refactoring of the current multi-level-4d-chess input validation process. We aim to make it an exported function, but leave it internal until we have a clearer plan for what we want that workflow to look like. Otherwise we commit to maintaining it now and have to do all the work of thinking about that workflow now.

Reply:
Issue: https://github.com/epiforecasts/scoringutils/issues/852